plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,842 @@
+            # [nparticles, 3] shape field array
+            dt = dt[tracked_mask, np.newaxis]
+
+            # Increment the tracked particles' time by dt
+            self.time[tracked_mask] += dt
+
+        self.x[tracked_mask], self.v[tracked_mask] = boris_push(
+            pos_tracked, vel_tracked, B, E, self.q, self.m, dt, inplace=False
+        )
+
+    @property
+    def on_any_grid(self):
+        """
+        Binary array for each particle indicating whether it is currently
+        on ANY grid.
+        """
+        return np.sum(self.on_grid, axis=-1) > 0
+
+    @property
+    def vmax(self):
+        """
+        Calculate the maximum velocity.
+        Used for determining the grid crossing maximum time step.
+        """
+        tracked_mask = self._tracked_particle_mask()
+
+        return np.max(np.linalg.norm(self.v[tracked_mask], axis=-1))
+
+    def _validate_run_inputs(
+        self, stop_condition, save_routine, dt, field_weighting: str
+    ):
+        if not isinstance(stop_condition, AbstractStopCondition):
+            raise TypeError("Please specify a valid stop condition.")
+
+        if not isinstance(save_routine, AbstractSaveRoutine):
+            raise TypeError("Please specify a valid save routine")
+
+        require_uniform_time = (
+            stop_condition.require_uniform_dt or save_routine.require_uniform_dt
+        )
+
+        # Will the simulation follow a uniform time step?
+        # This must be the case for certain stopping conditions and saving routines
+        self.is_uniform_time = isinstance(dt, u.Quantity) and isinstance(
+            dt.value, float
+        )
+
+        # Raise a ValueError if dt is required by stop condition or save routine but not specified
+        if require_uniform_time and not self.is_uniform_time:
+            raise ValueError(
+                "Please specify a uniform time step to use the simulation with this configuration!"
+            )
+
+        # Load and validate inputs
+        field_weightings = ["volume averaged", "nearest neighbor"]
+        if field_weighting in field_weightings:
+            self.field_weighting = field_weighting
+        else:
+            raise ValueError(
+                f"{field_weighting} is not a valid option for ",
+                "field_weighting. Valid choices are",
+                f"{field_weightings}",
+            )
+
+    def _tracked_particle_mask(self):
+        """
+        Calculates a boolean mask corresponding to particles that have not been stopped or removed.
+        """
+        return ~np.logical_or(np.isnan(self.x[:, 0]), np.isnan(self.v[:, 0]))
+
+    @property
+    def nparticles_tracked(self):
+        """Return the number of particles that don't have NaN position or velocity."""
+        return self._tracked_particle_mask().sum()
+
+    def run(
+        self,
+        stop_condition: AbstractStopCondition,
+        save_routine: Optional[AbstractSaveRoutine] = None,
```

Comment:
Here `save_routine` needs a default of class `AbstractSaveRoutine` for the case where the user doesn't want to save the data (i.e. they only care about the final state of the particles).